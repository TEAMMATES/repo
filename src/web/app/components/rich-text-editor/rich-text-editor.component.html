<div inViewport (inViewportAction)="renderEditor($event)">
  <editor class="editor" [ngClass]="{ 'editor-disabled': isDisabled }" *ngIf="render"
          [init]="init" [disabled]="isDisabled"
          [ngModel]="richText" (ngModelChange)="richTextChange.emit($event)" [attr.placeholder]="placeholderText">
  </editor>
  <div *ngIf="!isDisabled">
    {{richText ? RICH_TEXT_EDITOR_MAX_CHARACTER_LENGTH - richText.length + 7 : RICH_TEXT_EDITOR_MAX_CHARACTER_LENGTH}} characters left <br>
    {{richText ? RICH_TEXT_EDITOR_MAX_WORD_LENGTH - richText.split(' ').length : RICH_TEXT_EDITOR_MAX_WORD_LENGTH}} words left
  </div>
</div>
