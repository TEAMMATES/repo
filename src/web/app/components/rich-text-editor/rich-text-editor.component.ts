import { AfterViewInit, Component, EventEmitter, Input, OnDestroy, Output } from '@angular/core';
declare var tinymce: any;

/**
 * A rich text editor.
 */
@Component({
  selector: 'tm-rich-text-editor',
  templateUrl: './rich-text-editor.component.html',
  styleUrls: ['./rich-text-editor.component.scss'],
})
export class RichTextEditorComponent implements AfterViewInit, OnDestroy {

  @Input()
  isDisabled: boolean = false;

  @Input()
  isInlineMode: boolean = true;

  @Input()
  minHeightInPx: number = 120;

  @Input()
  placeholderText: string = '';

  @Input()
  richText: string = '';

  @Output()
  richTextChange: EventEmitter<string> = new EventEmitter();

  editor: any = '';

  // the argument passed to tinymce.init() in native JavaScript
  readonly init: any = {
    skin_url: '/tinymce/skins/ui/oxide',
    resize: false,
    selector: '.editor',
    fontsize_formats: '8pt 9pt 10pt 11pt 12pt 14pt 16pt 18pt 20pt 24pt 26pt 28pt 36pt 48pt 72pt',
    font_formats: 'Andale Mono=andale mono,times;'
        + 'Arial=arial,helvetica,sans-serif;'
        + 'Arial Black=arial black,avant garde;'
        + 'Book Antiqua=book antiqua,palatino;'
        + 'Comic Sans MS=comic sans ms,sans-serif;'
        + 'Courier New=courier new,courier;'
        + 'Georgia=georgia,palatino;'
        + 'Helvetica=helvetica;'
        + 'Impact=impact,chicago;'
        + 'Symbol=symbol;'
        + 'Tahoma=tahoma,arial,helvetica,sans-serif;'
        + 'Terminal=terminal,monaco;'
        + 'Times New Roman=times new roman,times;'
        + 'Trebuchet MS=trebuchet ms,geneva;'
        + 'Verdana=verdana,geneva;'
        + 'Webdings=webdings;'
        + 'Wingdings=wingdings,zapf dingbats',
    relative_urls: false,
    convert_urls: false,
    remove_linebreaks: false,
    plugins: [
      'placeholder',
      'advlist autolink lists link image charmap print hr anchor',
      'searchreplace wordcount visualblocks visualchars code fullscreen',
      'insertdatetime nonbreaking save table directionality',
      'emoticons paste textpattern',
    ],

    toolbar1: 'insertfile undo redo | styleselect | bold italic underline '
        + '| alignleft aligncenter alignright alignjustify '
        + '| bullist numlist outdent indent | link image',
    toolbar2: 'print | forecolor backcolor | fontsizeselect fontselect | emoticons | fullscreen',
    setup: (editor: any) => {
      this.editor = editor;
      editor.on('keyup change', () => {
        const content = editor.getContent();
        this.richTextChange.emit(content);
      });
    },
    init_instance_callback: (editor: any) => editor.setContent(this.richText),
  };

  constructor() { }

  ngAfterViewInit(): void {
    tinymce.init({ ...this.init, readonly: this.isDisabled, inline: this.isInlineMode });
  }

  ngOnDestroy() {
    tinymce.remove(this.editor);
  }
}
