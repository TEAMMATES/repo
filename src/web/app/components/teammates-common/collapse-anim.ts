import { animate, AnimationTriggerMetadata, state, style, transition, trigger } from '@angular/animations';

/**
 * Directive for collapsing of *ngIf columns.
 */
export const collapseAnim: AnimationTriggerMetadata = trigger('collapseAnim', [
  state('*', style({ overflow: 'hidden' })),
  state('void', style({ overflow: 'hidden' })),
  transition(':leave', [
    style({ height: '*' }),
    animate('300ms ease-in-out', style({ height: 0, opacity: 0 })),
  ]),
  transition(':enter', [
    style({ height: '0' }),
    animate('300ms ease-in-out', style({ height: '*', opacity: 1 })),
  ]),
]);
