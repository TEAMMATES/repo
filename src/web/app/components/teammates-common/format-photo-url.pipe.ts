import { Pipe, PipeTransform } from '@angular/core';
import { environment } from "../../../environments/environment";

@Pipe({
  name: 'formatPhotoUrl'
})
export class FormatPhotoUrlPipe implements PipeTransform {

  transform(courseId: string, email: string): string {
    return `${environment.backendUrl}/webapi/student/profilePic?courseid=${courseId}&studentemail=${email}`;
  }
}
