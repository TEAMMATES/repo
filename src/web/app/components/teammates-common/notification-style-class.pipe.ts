import { Pipe, PipeTransform } from '@angular/core';
import { NotificationStyle } from '../../../types/api-output';
import { notificationStyleConfigMap } from '../../../types/notification-style-config';

/**
 * Pipe to handle the transformation of an NotificationStyle to a css class.
 */
@Pipe({
  name: 'notificationStyleClass',
})
export class NotificationStyleClassPipe implements PipeTransform {

  transform(style: NotificationStyle): string {
    return `bg-${style.toLowerCase()} ${notificationStyleConfigMap[style].font}`;
  }

}
