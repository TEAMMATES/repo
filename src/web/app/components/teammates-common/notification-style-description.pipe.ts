import { Pipe, PipeTransform } from '@angular/core';
import { NotificationStyle } from '../../../types/api-output';
import { notificationStyleConfigMap } from '../../../types/notification-style-config';

/**
 * Pipe to handle the transformation of an NotificationStyle to a string description.
 */
@Pipe({
  name: 'notificationStyleDescription',
})
export class NotificationStyleDescriptionPipe implements PipeTransform {

  transform(style: NotificationStyle): string {
    return notificationStyleConfigMap[style].description;
  }

}
