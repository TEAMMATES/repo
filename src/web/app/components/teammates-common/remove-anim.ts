import { animate, AnimationTriggerMetadata, style, transition, trigger } from '@angular/animations';

/**
 * Directive for removing *ngFor items.
 */
export const removeAnim: AnimationTriggerMetadata = trigger('removeAnim', [
  transition(':leave', [
    style({ height: '*', opacity: 1, overflow: 'hidden' }),
    animate('300ms ease-in-out', style({ height: 0, opacity: 0 })),
  ]),
]);
