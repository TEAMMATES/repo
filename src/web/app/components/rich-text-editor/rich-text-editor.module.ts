import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RichTextEditorComponent } from './rich-text-editor.component';

/**
 * Module for a rich text editor.
 */
@NgModule({
  declarations: [RichTextEditorComponent],
  imports: [
    CommonModule,
    FormsModule,
  ],
  exports: [
    RichTextEditorComponent,
  ],
})
export class RichTextEditorModule { }
