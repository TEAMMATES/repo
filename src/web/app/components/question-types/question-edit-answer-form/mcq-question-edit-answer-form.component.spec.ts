import { DebugElement } from '@angular/core';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { FeedbackMcqQuestionDetails } from 'src/web/types/api-output';
import { RichTextEditorModule } from '../../rich-text-editor/rich-text-editor.module';
import { TeammatesCommonModule } from '../../teammates-common/teammates-common.module';
import { McqQuestionEditAnswerFormComponent } from './mcq-question-edit-answer-form.component';

describe('McqQuestionEditAnswerFormComponent', () => {
  let component: McqQuestionEditAnswerFormComponent;
  let fixture: ComponentFixture<McqQuestionEditAnswerFormComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [McqQuestionEditAnswerFormComponent],
      imports: [
        FormsModule,
        RichTextEditorModule,
        TeammatesCommonModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(McqQuestionEditAnswerFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when questionDropdownEnabled is updated', () => {
    component.questionDetails.questionDropdownEnabled = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should display all MCQ choices for question dropdown disabled', () => {
    component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
    fixture.detectChanges();

    const radioGroup: DebugElement[] = fixture.debugElement.queryAll(By.css('#radioOptionSpan'));
    const mcqChoices: FeedbackMcqQuestionDetails['mcqChoices'] = component.questionDetails.mcqChoices;
    const radioSpans: HTMLSpanElement[] = [];

    radioGroup.forEach((de: DebugElement) => radioSpans.push(de.nativeElement));

    for (let i = 0; i < radioSpans.length; i += 1) {
      expect(radioSpans[i].innerHTML).toStrictEqual(mcqChoices[i]);
    }
  });

  it('should display all MCQ choices for question dropdown enabled', () => {
    component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
    component.questionDetails.questionDropdownEnabled = true;
    fixture.detectChanges();

    const select: HTMLSelectElement = fixture.debugElement.query(By.css('#dropdown-option-select')).nativeElement;

    // Removes placeholder option
    select.options.remove(0);

    const mcqChoices: FeedbackMcqQuestionDetails['mcqChoices'] = component.questionDetails.mcqChoices;

    for (let i = 0; i < select.options.length; i += 1) {
      expect(select.options[i].text).toStrictEqual(mcqChoices[i]);
    }
  });
});
