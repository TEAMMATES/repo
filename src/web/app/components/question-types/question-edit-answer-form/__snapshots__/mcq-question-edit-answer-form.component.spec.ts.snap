// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`McqQuestionEditAnswerFormComponent should snap when questionDropdownEnabled is updated 1`] = `
<tm-mcq-question-edit-answer-form
  cssRefresh={[Function EventEmitter_]}
  id=""
  inputTextBoxOther="undefined"
  isDisabled="false"
  isMcqOptionSelected={[Function Array]}
  isOtherTicked="false"
  questionDetails={[Function Object]}
  responseDetails={[Function Object]}
  responseDetailsChange={[Function EventEmitter_]}
>
  <tr>
    <td>
      <div>
        <select
          class="form-control margin-bottom-8px"
          id="dropdown-option-select"
        >
          <option
            disabled=""
            selected=""
          />
        </select>
      </div>
    </td>
  </tr>
</tm-mcq-question-edit-answer-form>
`;
